loadfile/loadfile.go

```diff
@@ -7,41 +7,168 @@ import (
+// file contents. rootDir is the context directory (root directory) in
+// which the files can be found. 'files' is a mapping of the desired file
+// key to a relative or absolute file path.
+func NewFileCacheUsingContext(rootDir string, files map[string]string) (FileCache, error) {
 	absDir, err := filepath.Abs(rootDir)
 	if err != nil {
 		return nil, fmt.Errorf("error determining context directory absolute path %s (%w)", rootDir, err)
 	}
-	requiredFilesAbs := map[string]string{}
-	for key, f := range requiredFiles {
+	filesAbsPaths := map[string]ContextFile{}
+	for key, f := range files {
 		abspath := f
 		if !filepath.IsAbs(f) {
 			abspath = filepath.Join(absDir, f)
 		}
-		requiredFilesAbs[key] = abspath
+		filesAbsPaths[key] = ContextFile{
+			ID:           f,
+			AbsolutePath: abspath,
+		}
+	}
+	return &fileCache{
+		rootDir: absDir, files: filesAbsPaths}, nil
+}
+
+// NewFileCache creates a file cache where the fileContents is
+// a mapping of paths to file contents. The keys in fileContents
+// become the keys in Files, and the absolute paths and ID for
+// each context file.
+func NewFileCache(rootDir string, fileContents map[string][]byte) FileCache {
+	files := map[string]ContextFile{}
+	for key, content := range fileContents {
+		files[key] = ContextFile{
+			ID:           key,
+			AbsolutePath: key,
+			Content:      content,
+		}
+	}
+	return &fileCache{
+		rootDir: rootDir,
+		files:   files,
+	}
+}
+
+// LoadContext reads the content of each context file into Content.
+func (fc *fileCache) LoadContext() error {
+	result := map[string]ContextFile{}
+	var err error
+	for key, cf := range fc.files {
+		absPath := cf.AbsolutePath
+		fileData, err := os.ReadFile(filepath.Clean(absPath))
+		if err != nil {
+			return fmt.Errorf("error reading file %s (%w)", absPath, err)
+		}
+		result[key] = ContextFile{
+			ID:           cf.ID,
+			AbsolutePath: cf.AbsolutePath,
+			Content:      fileData,
+		}
+	}
+	fc.files = result
+	return err
+}
+
+// RootDir returns the root directory used by files with
+// relative file paths.
+func (fc *fileCache) RootDir() string {
+	return fc.rootDir
+}
+
+// GetByKey asks if the file cache contains the given fileKey.
+func (fc *fileCache) GetByKey(fileKey string) (*ContextFile, bool) {
+	cf, ok := fc.files[fileKey]
+	return &cf, ok
+}
+
+// AbsPathByKey returns the absolute file path of a given file key,
+// if it exists in the file cache, nil otherwise.
+func (fc *fileCache) AbsPathByKey(fileKey string) *string {
+	cf, ok := fc.GetByKey(fileKey)
+	if !ok {
+		return nil
+	}
+	return &cf.AbsolutePath
+}
+
+// ContentByKey returns the file content of a given file key, if it
+// exists in the file cache, nil otherwise.
+func (fc *fileCache) ContentByKey(fileKey string) []byte {
+	cf, ok := fc.GetByKey(fileKey)
+	if !ok {
+		return nil
+	}
+	return cf.Content
+}
+
+// Contents returns a mapping of the file cache's file keys to file Content.
+func (fc *fileCache) Contents() map[string][]byte {
```

Comment:
I realize that we're working to minimize changes and maximize compatibility, but the fact that we have to actively convert back and forth between `map[string][]byte` and `fileCache`/`ContextFile` suggests that we should have a different implementation.

For example, perhaps the `fileCache` should contain two separate maps, one for the paths and one for the contents.  These could be constructed independently, and, we wouldn't have to convert back in `Contents()` and `Abspaths()` -- we could just return the appropriate already-constructed map.

Reply:
I preferred composing the information together into one object, so that I didn't have to coordinate updates between two hash maps.

Reply:
Yeah, and doing so ensures that there is no diversion in the key-space.  But, taking the contents map, converting it to a map of `ContextFile`'s, only to convert it _back_ is really unsatisfying.  The fact that someone is going to want the map of contexts later is justification enough (to me) to retain that map in the `FileContext`.

Reply:
It felt unsatisfying to me as well. The design is a compromise in part to put off deciding how we want the rest of the engine to access `Content`. Right now, I would rather `Content` be accessed through the `FileCache` interface, but that is not a necessary part to refactor to get to the sub-workflow bug fix.

Reply:
And, I'm hoping that this will be reworked when the code is refactored.